src/libs/ck-libs/ampi/ddt.h

```diff
@@ -443,18 +445,17 @@ class CkDDT_Struct final : public CkDDT_DataType
  * This class maintains the table of all datatypes (predefined and user-defined).
  *
  * predefinedTypeTable - a reference to a const array declared as a static global variable
- *                       (to minimize per-rank memory fooprint), which holds the CkDDT_DataType
+ *                       (to minimize per-rank memory footprint), which holds the CkDDT_DataType
  *                       object pointers for all predefined types.
  * userTypeTable - a vector that holds the CkDDT_DataType object pointers for all user-defined types
- * types - used to identify which CkDDT_DataType derived class a type object really is,
- *         for PUPing the userTypeTable
+ * freeTypes - an ordered set of freed slot indexes in the userTypeTable, available for reuse.
  */
 class CkDDT
 {
  private:
   const std::array<const CkDDT_DataType *, AMPI_MAX_PREDEFINED_TYPE+1>& predefinedTypeTable;
   std::vector<CkDDT_DataType *> userTypeTable;
-  std::vector<int> types;
+  std::set<int> freeTypes; // TODO: replace with a priority queue structure that can be serialized trivially
```

Comment:
Out of curiosity, why is a priority queue the intended data structure here? Could we use an `unordered_map` +  int (as an index)?

Reply:
ie, something like 
```c++
std::unordered_map<int, CkDDT_DataType *> userTypeTable;
int lastIndex;

// in insertType()
userTypeTable[++lastIndex] = ptr;
return lastIndex;
```

Reply:
How would changing the type of userTypeTable from a vector to an unordered_map solve the problem of knowing which indexes are empty and available for reused?

freeTypes is a priority queue so that AMPI can fill in empty slots from least to greatest, so that types freed at the end of the list have the greatest change of being able to shrink userTypeTable as a whole.

Reply:
> How would changing the type of userTypeTable from a vector to an unordered_map solve the problem of knowing which indexes are empty and available for reuse?

Do we need to know which indices can be reused (ie, can't we just increase the indices)?